assembly/src/assembler/mod.rs

```diff
@@ -300,27 +304,35 @@ impl Assembler {
         &self,
         proc_id: &ProcedureId,
         context: &mut AssemblyContext,
-    ) -> Result<&Procedure, AssemblyError> {
-        // if the procedure is already in the procedure cache, return it
-        if let Some(p) = self.proc_cache.get(proc_id) {
-            return Ok(p);
+    ) -> Result<impl Deref<Target = Procedure> + '_, AssemblyError> {
+        if !self.proc_cache.borrow().contains_key(proc_id) {
+            // if procedure is not in cache, try to get its module and compile it
+            let module = self
+                .module_provider
+                .get_module(proc_id)
+                .ok_or_else(|| AssemblyError::imported_proc_module_not_found(proc_id))?;
+            self.compile_module(&module, module.path(), context)?;
+
+            // if the procedure is still not in cache, then there was some error
+            if !self.proc_cache.borrow().contains_key(proc_id) {
+                return Err(AssemblyError::imported_proc_not_found_in_module(
+                    proc_id,
+                    module.path(),
+                ));
+            }
         }
 
-        // otherwise, get the module to which the procedure belongs and compile the entire module;
-        // this will add all procedures exported from the module to the procedure cache
-        let module = self
-            .module_provider
-            .get_module(proc_id)
-            .ok_or_else(|| AssemblyError::imported_proc_module_not_found(proc_id))?;
-        self.compile_module(&module, module.path(), context)?;
-
-        // then, get the procedure out of the procedure cache and return; if the procedure
-        // cannot be found in the cache, it is possible that the procedure was not in the
-        // module returned from the module provider
-        let proc = self.proc_cache.get(proc_id).ok_or_else(|| {
-            AssemblyError::imported_proc_not_found_in_module(proc_id, module.path())
-        })?;
-        Ok(proc)
+        // the procedure was checked to be in the map; we can safely unwrap it
+        Ok(Ref::map(self.proc_cache.borrow(), |cache| {
+            cache
+                .get(proc_id)
+                .map(|proc| {
+                    // this map is reduced/removed in release mod
+                    debug_assert!(proc.is_export(), "not imported procedure");
+                    proc
+                })
+                .expect("proc id was checked in the cache. this is a bug in the compiler")
```

Comment:
I think this could be also a bug in the module provider (i.e., it reported that a procedure was in a specific module, but when the module was compiled, the procedure was not there).

Reply:
The faulty part is intended to point at this block, when a module is effectively returned from the provider, compiled, but not added to the cache

https://github.com/maticnetwork/miden/blob/a4adb344d8278fde1161eac52ec46ddb644191e8/assembly/src/assembler/mod.rs#L314-L317

I'll rephrase the message.